mitiq/cdr/clifford_training_data.py

```diff
@@ -0,0 +1,516 @@
+    rz_non_cliff = angles[mask_non_cliff]
+    return len(rz_non_cliff)
+
+
+def _is_clifford_angle(ang: float, tol: float = 10 ** -5,) -> bool:
+    """Function to check if a given angle is Clifford.
+    Args:
+        ang: rotation angle in the Rz gate.
+    Returns:
+        bool: True / False for Clifford or not.
+    """
+    ang = ang % (2 * np.pi)
+    closest_clifford_angle = _closest_clifford(ang)
+    if abs(closest_clifford_angle - ang) < tol:
+        return True
+    else:
+        return False
+
+
+# Vectorize function so it can take arrays of angles as its input.
+_is_clifford_angle = np.vectorize(_is_clifford_angle)
+
+
+def _closest_clifford(ang: float) -> float:
+    """Function to take angle and return the nearest Clifford angle note the
+       usage of this function is vectorized so it takes and returns arrays.
+
+    Args:
+        ang: angle in Rz gate.
+
+    Returns:
+        Clifford angle: closest clifford angle.
+    """
+    ang = ang % (2 * np.pi)
+    ang_scaled = ang / (np.pi / 2)
+    # if just one min value, return the corresponding nearest cliff.
+    if (
+        abs((ang_scaled / 0.5) - 1) > 10 ** (-6)
+        and abs((ang_scaled / 0.5) - 3) > 10 ** (-6)
+        and (abs((ang_scaled / 0.5) - 5) > 10 ** (-6))
+    ):
+        index = int(np.round(ang_scaled)) % 4
+        return CLIFFORD_ANGLES[index]
+    # if two min values (ie two cliff gates equidistant) randomly choose the
+    # cliff gate to return.
+    else:
+        index_list = [ang_scaled - 0.5, ang_scaled + 0.5]
+        index = int(choice(index_list))
+        return CLIFFORD_ANGLES[index]
+
+
+# vectorize so function can take array of angles.
+_closest_clifford = np.vectorize(_closest_clifford)
+
+
+def _random_clifford(ang: float, random_state: np.random.RandomState) -> float:
+    """Function to take angle and return the random Clifford angle note the
+       usage of this function is vectorized so it takes and returns arrays.
+
+    Args:
+        ang: angle in Rz gate.
+
+    Returns:
+        Clifford angle: closest clifford angle.
+    """
+    random_index = random_state.randint(0, 3)
+    clifford_angle = CLIFFORD_ANGLES[random_index]
+    return clifford_angle
+
+
+# vectorize so function can take array:
+_random_clifford = np.vectorize(_random_clifford)
+
+
+def _angle_to_probabilities(angle: float, sigma: float) -> float:
+    """Function to return probability disribtuion based on distance from
+       angles to Clifford gates.
+
+    Args:
+        angle: angle to form probability distribution.
+    Returns:
+        discrete value of probability distribution calucalted from
+        Prob_project = exp(-(dist/sigma)^2) where dist = sum(dists) is the
+        sum of distances from each Clifford gate.
+    """
+    angle = angle % (2 * np.pi)
+    S = np.array([[1, 0.0], [0.0, 1j]])
+    Rz = np.array([[1, 0.0], [0.0, np.exp(angle * 1j)]])
+    dists = []
+    for i in range(4):
+        if i == 0:
+            i = 4
```

Comment:
why start at 0 then?

Reply:
Because I want the order to correspond to the order of the tuple CLIFFORD_ANGLES. I could reorder that list as well I suppose, but that does not appear immediately more elegant to me...